Blog post to introduce the Temporal Swift SDK (#1221)
apple/swift-org-website: _posts/2025-11-10-swift-temporal-sdk.md

---
layout: new-layouts/post
published: true
date: 2025-11-10 10:00:00
title: "Introducing Temporal Swift SDK: Building durable and reliable workflows"
author: [FranzBusch]
category: "Developer Tools"
---

[Temporal Swift SDK](https://github.com/apple/swift-temporal-sdk), a Swift package for building reliable distributed systems with durable execution, is now available as an open source project.

Building reliable distributed systems requires handling failures gracefully, coordinating complex workflows across multiple services, and ensuring long-running processes complete successfully. With the Temporal Swift SDK, you can now build fault-tolerant applications using Swift's modern concurrency features. The Temporal Swift SDK enables you to:

> - **Build resilient workflows that survive infrastructure failures**, allowing your applications to recover automatically from crashes, network issues, or service restarts without losing state.

Reviewer: ```suggestion
- Build resilient workflows that survive infrastructure failures.
```

- **Orchestrate complex distributed systems with type-safe APIs**, leveraging Swift's strong type system to catch errors at build time rather than runtime.
- **Integrate seamlessly with Swift Concurrency**, using familiar async/await patterns and Structured Concurrency to build maintainable workflow code.

> Temporal has also published a [blog post](http://temporal.io/blog/announcing-the-swift-temporal-sdk) detailing the introduction of the Temporal Swift SDK highlighting the unique advantages this brings to Swift developers building distributed systems.

Reviewer: +1, I agree this may make sense in the Temporal section.

> Temporal has also published a [blog post](http://temporal.io/blog/announcing-the-swift-temporal-sdk) detailing the introduction of the Temporal Swift SDK highlighting the unique advantages this brings to Swift developers building distributed systems.

Reviewer: ```suggestion
```

> This new Temporal Swift SDK fills a critical gap in the Swift ecosystem, enabling developers to write reliable and durable workflows with native Swift APIs and modern concurrency patterns. Before this SDK, Swift developers had to resort to more traditional approaches for building fault-tolerant workflows, such as implementing custom state machines on top of database systems or message queuing services, solutions that often required significant boilerplate code and careful handling of edge cases that Temporal manages automatically.

Reviewer: ```suggestion
## Durable workflows in Swift with Temporal

This new Temporal Swift SDK fills a critical gap in the Swift ecosystem, enabling developers to write reliable and durable workflows with native Swift APIs and modern concurrency patterns. Before this SDK, Swift developers had to resort to more traditional approaches for building fault-tolerant workflows, such as implementing custom state machines on top of database systems or message queuing services, solutions that often required significant boilerplate code and careful handling of edge cases that Temporal manages automatically.
```

The design and implementation of the Temporal Swift SDK has been informed and validated by multiple production use-cases, ensuring that the SDK meets real-world requirements for performance, reliability, and developer experience.

> The Temporal Swift SDK can be used across a wide range of real-world scenarios where reliability and coordination are paramount, including:

Reviewer: I would suggest that this paragraph and list items are either:

1. replaced with a code example illustrating a simple example of one of these uses
2. reframed and the list items framed as technical challenges vs business problems. For example, Temporal is useful for services that need to be highly scalable, can't lose data, etc.
3. remove it to simplify things overall

Reviewer: I would also recommend elaborating a bit more Swift language features -- how the SDK takes advantage of those features and why it's a good choice.

---
layout: new-layouts/post
published: true
date: 2025-11-10 10:00:00
title: "Introducing Temporal Swift SDK: Building durable and reliable workflows"
author: [FranzBusch]
category: "Developer Tools"
featured-image:
  url: '/assets/images/swift-temporal-sdk-blog/blog-hero@2x.png'
  alt: 'Featured blog image with the diagram of a Temporal workflow'
featured-image-dark:
  url: '/assets/images/swift-temporal-sdk-blog/blog-hero-dark@2x.png'
  alt: 'Featured blog image with the diagram of a Temporal workflow'
---

The [Temporal Swift SDK](https://github.com/apple/swift-temporal-sdk) is now available as an open source project. This Swift package enables you to build reliable distributed systems using Temporal’s durable execution platform.

> Building reliable distributed systems requires handling failures gracefully, coordinating complex workflows across multiple services, and ensuring long-running processes complete successfully. Rather than develop these resiliency features into every application or service you develop, Temporal offers the pattern of a __workflow__. Workflows encapsulate your code so it runs durably and handles many common failure scenarios.

Reviewer 2: ```suggestion
Building reliable distributed systems requires handling failures gracefully, coordinating complex actions across multiple services, and ensuring long-running processes complete successfully. Rather than develop these resiliency features into every application or service you develop, an alternative approach is to think of your system as a __workflow__ of messages and signals that pass between services. Workflows encapsulate your code so it runs durably and handles many common failure scenarios.
```

Reviewer: @timsneath Great point re: use of workflows too early; I'd suggest "complex processes" instead of "complex actions".

The expanded definition of workflow may add confusion, so I'm exploring simplified ways of making this change.

Reviewer 2: But isn't "process" is confusing here? Later in the sentence we mean "system process", I think, but here I think you're using the word to mean "business process"?

Reviewer: @timsneath good point, I'll use your earlier suggestion re: actions.

> Temporal is widely used for workflow orchestration across many languages. With the release of the SDK, Temporal is now available for Swift developers building production cloud services.

Reviewer: ```suggestion
Temporal is widely used for workflow orchestration across many languages. With the release of the SDK, Temporal is now available for Swift developers building production cloud services.

An [announcement post](https://temporal.io/blog/announcing-the-swift-temporal-sdk) is also available on the Temporal blog with more information for their developer community.
```

## Writing durable workflows in Swift

Durable workflows run to completion even when infrastructure fails. If your server crashes mid-execution, Temporal automatically resumes the workflow from where it left off with no lost state or manual intervention required. By bringing this capability to the Swift ecosystem, developers can focus on application logic while Temporal handles state management, retries, and recovery.

The SDK provides a seamless Swift developer experience by:

> - Using familiar async/await patterns and Structured Concurrency to build maintainable workflow code.

Reviewer 2: ```suggestion
- Using familiar async/await patterns and structured concurrency to build maintainable workflow code.
```

We don't title case "structured concurrency". See, for example, https://developer.apple.com/videos/play/wwdc2023/10170

---
layout: new-layouts/post
published: true

> date: 2025-11-10 10:00:00

Reviewer: ```suggestion
date: 2025-11-10 13:40:00
```

title: "Introducing Temporal Swift SDK: Building durable and reliable workflows"
author: [FranzBusch]
category: "Developer Tools"
featured-image:
  url: '/assets/images/swift-temporal-sdk-blog/blog-hero@2x.png'
  alt: 'Featured blog image with the diagram of a Temporal workflow'
featured-image-dark:
  url: '/assets/images/swift-temporal-sdk-blog/blog-hero-dark@2x.png'
  alt: 'Featured blog image with the diagram of a Temporal workflow'
---

The [Temporal Swift SDK](https://github.com/apple/swift-temporal-sdk) is now available as an open source project.

Building reliable distributed systems requires handling failures gracefully, coordinating complex actions across multiple services, and ensuring long-running processes complete successfully. Rather than develop these resiliency features into every application or service you develop, an alternative approach is to use __workflows__. Workflows encapsulate your code so it runs durably and handles many common failure scenarios.

Temporal is widely used for workflow orchestration across many languages. With the release of the SDK, Temporal is now available for Swift developers building production cloud services.

> An [announcement post](https://temporal.io/blog/announcing-the-swift-temporal-sdk) is also available on the Temporal blog with more information for their developer community.

Reviewer: ```suggestion
```

With the Temporal Swift SDK, you can build resilient workflows that survive infrastructure failures. The SDK brings another essential tool for developers that build highly reliable and scalable production cloud services.

## Writing durable workflows in Swift

The Temporal Swift SDK enables developers to write reliable and durable workflows. Previously, Swift developers had to resort to more traditional approaches for building fault-tolerant workflows, such as implementing custom state machines on top of database systems or message queuing services, solutions that often required significant boilerplate code and careful handling of edge cases that Temporal manages automatically.

The SDK provides a seamless Swift developer experience by:

- Using familiar async/await patterns and Structured Concurrency to build maintainable workflow code.
- Leveraging Swift's strong type system to catch errors at build time rather than runtime.
- Providing macros to reduce boilerplate when authoring workflows.

> The design and implementation of the Temporal Swift SDK has been informed and validated by multiple production use-cases, ensuring that the SDK meets real-world requirements for performance, reliability, and developer experience. 

Reviewer: It may be worth removing this paragraph.

> 

Reviewer: ```suggestion
```

Reviewer: Delete newline if the paragraph above is also delete.

The Temporal Swift SDK can be used across a wide range of real-world scenarios where reliability and coordination are paramount, including:

> 

Reviewer: ```suggestion
```

> - E-commerce and payment processing where multi-step payment processing requires automatic retry and rollback capabilities.

Reviewer: ```suggestion
```

> - Data processing & ETL where you may be performing large-scale data transformation pipelines.

Reviewer: ```suggestion
```

The Temporal Swift SDK can be used across a wide range of real-world scenarios where reliability and coordination are paramount, including:

- E-commerce and payment processing where multi-step payment processing requires automatic retry and rollback capabilities.
- Data processing & ETL where you may be performing large-scale data transformation pipelines.
- Business process automation with approval workflows and human-in-the-loop interactions.
- Monitoring and operations with scheduled maintenance and cleanup tasks.

## What is Temporal and why does it matter?

[Temporal](https://temporal.io) is an open source platform for building reliable distributed applications. At its core is the concept of durable execution, your code runs to completion even in the face of infrastructure failures. When a worker crashes or restarts, Temporal automatically resumes your workflow from where it left off, without requiring you to write complex retry logic or state management code. This is achieved through Temporal's architecture, which separates workflow orchestration from actual work execution:

> - **Workflows** define the overall business logic and coordination. They describe the sequence of steps, decision points, and error handling for a process. Workflows must be deterministic given the same inputs and history, they must always make the same decisions.

Reviewer: ```suggestion
- **Workflows** define the overall business logic and coordination. They describe the sequence of steps, decision points, and error handling for a process. Workflows must be deterministic. Given the same inputs and history, they must always make the same decisions.
```

- **Activities** perform the actual work, such as calling external APIs, processing data, or interacting with databases. Activities should be idempotent, meaning they can be safely retried without causing unintended side effects.

Modern distributed systems face common challenges: coordinating multiple services, handling partial failures, ensuring consistency across operations, and managing long-running processes. Traditional approaches require building custom retry logic, state machines, and coordination mechanisms. Temporal provides a platform that handles these concerns, allowing you to focus on your business logic.

Temporal has also published a [blog post](https://temporal.io/blog/announcing-the-swift-temporal-sdk) detailing the introduction of the Temporal Swift SDK highlighting the unique advantages this brings to Swift developers building distributed systems.

> 

Reviewer: ```suggestion
```

## Getting started

To get started with the Temporal Swift SDK, explore its [documentation](https://swiftpackageindex.com/apple/swift-temporal-sdk/main/documentation/temporal) which provides detailed guides for implementing workflows and activities. The repository also includes a rich collection of [example projects](https://github.com/apple/swift-temporal-sdk/tree/main/Examples), demonstrating the SDK's capabilities across different use cases from simple task orchestration to complex multi-step business processes. For a deeper understanding of Temporal's core concepts and architectural patterns, be sure to check out the [general Temporal documentation](https://docs.temporal.io/), which provides valuable context for building robust distributed systems.

## Community and feedback

Temporal Swift SDK is an open source project and we're eager to hear from the Swift community. Whether you're building microservices, coordinating long-running processes, or simply curious about durable execution, we'd love to know how the Temporal Swift SDK works for you. The SDK provides a powerful foundation for building reliable distributed systems in Swift.

> The project is actively developed and we welcome contributions, bug reports, and feature requests. As with any distributed systems tool, different applications and environments present unique challenges, and community feedback is essential for improving the SDK's effectiveness across diverse Swift codebases.

Reviewer: ```suggestion
The project is actively developed and we welcome contributions, bug reports, and feature requests. Ready to start building reliable distributed systems? Visit the [Temporal Swift SDK repository](https://github.com/apple/swift-temporal-sdk) to get started.
```

> Ready to start building reliable distributed systems? Visit the [Temporal Swift SDK repository](https://github.com/apple/swift-temporal-sdk) to get started.

Reviewer: ```suggestion
```
